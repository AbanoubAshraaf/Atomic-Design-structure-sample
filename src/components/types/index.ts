
export interface BuyflowProps {
    productId: ProductIds;
}

export enum ProductIds {
    devIns = "dev_ins",
}

export const PRODUCT_IDS_TO_NAMES = {
    [ProductIds.devIns]: "Developer Insurance",
};

export interface CollectedData {
    email: string;
    age: number;
}

export interface StepProps {
    onClick: (field: keyof CollectedData, value: string | number) => void;
}

export interface SummaryStepProps {
    collectedData: CollectedData;
}