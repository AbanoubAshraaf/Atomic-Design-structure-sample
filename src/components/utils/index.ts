export const isValidAge = (age:number): boolean => {
    return (
      age > 0 && // Age must be positive
      !isNaN(age) && // Age must be a valid number
      Number.isInteger(age) // Age must be an integer
    );
  };
  