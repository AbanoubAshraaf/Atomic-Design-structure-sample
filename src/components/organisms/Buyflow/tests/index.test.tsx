import React from "react";
import { render, fireEvent } from "@testing-library/react";
import { ProductIds } from "@/components/types";
import BuyFlow from "..";
import { mockValidUserData } from "@/components/test/mocks";
import { createSnapShot } from "@/components/test/utils";


describe("BuyFlow Component", () => {
  test("display the correct product name", () => {
    const { container, getByText } = render(<BuyFlow productId={ProductIds.developerIns} />);
    expect(getByText("Buying Developer Insurance")).toBeInTheDocument();

    createSnapShot(container);
  });

  test("render email step initially", () => {
    const { getByTestId } = render(<BuyFlow productId={ProductIds.developerIns} />);
    expect(getByTestId("email")).toBeInTheDocument();

  });

  test("completes the buyflow process properly", () => {
    const { container, getByTestId, getByText } = render(<BuyFlow productId={ProductIds.developerIns} />);

    // Initial email step
    fireEvent.change(getByTestId("email"), { target: { value: mockValidUserData.email } });
    fireEvent.click(getByText("Next"));

    // Transition to age step
    fireEvent.change(getByTestId("age"), { target: { value: mockValidUserData.age } });
    fireEvent.click(getByText("Buy Now"));

    // Verify collected data in the summary step
    expect(getByText("Email: test@getsafe.com")).toBeInTheDocument();
    expect(getByText("Age: 20")).toBeInTheDocument();
    expect(getByText("Purchase")).toBeInTheDocument();

    createSnapShot(container);

  });
});
