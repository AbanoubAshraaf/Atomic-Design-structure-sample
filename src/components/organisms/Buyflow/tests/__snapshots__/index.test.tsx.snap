// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BuyFlow Component completes the buyflow process properly 1`] = `
<div>
  <div
    class="flex flex-col items-center"
  >
    <h4
      class="mb-6 text-2xl"
    >
      Buying 
      Developer Insurance
    </h4>
    <div
      class="flex justify-center w-60"
    >
      <div
        class="flex flex-col gap-4 w-full"
      >
        <div
          class="flex flex-col gap-2 items-start"
        >
          <div>
            Email: 
            test@getsafe.com
          </div>
          <div>
            Age: 
            20
          </div>
        </div>
        <a
          class="btn btn-primary"
          href="/purchased=dev_ins"
        >
          Purchase
        </a>
      </div>
    </div>
  </div>
</div>
`;

exports[`BuyFlow Component displays the correct product name 1`] = `
<div>
  <div
    class="flex flex-col items-center"
  >
    <h4
      class="mb-6 text-2xl"
    >
      Buying 
      Developer Insurance
    </h4>
    <div
      class="flex justify-center w-60"
    >
      <div
        class="flex flex-col gap-4 w-full"
      >
        <div
          class="flex flex-col gap-2 items-start"
        >
          <label>
            Email
            :
          </label>
          <input
            class="input input-bordered input-primary w-full max-w-xs"
            data-testid="email"
            type="email"
            value=""
          />
        </div>
        <button
          class="btn btn-primary"
        >
          Next
        </button>
      </div>
    </div>
  </div>
</div>
`;

exports[`BuyFlow Component renders email step initially 1`] = `
<div>
  <div
    class="flex flex-col items-center"
  >
    <h4
      class="mb-6 text-2xl"
    >
      Buying 
      Developer Insurance
    </h4>
    <div
      class="flex justify-center w-60"
    >
      <div
        class="flex flex-col gap-4 w-full"
      >
        <div
          class="flex flex-col gap-2 items-start"
        >
          <label>
            Email
            :
          </label>
          <input
            class="input input-bordered input-primary w-full max-w-xs"
            data-testid="email"
            type="email"
            value=""
          />
        </div>
        <button
          class="btn btn-primary"
        >
          Next
        </button>
      </div>
    </div>
  </div>
</div>
`;
